include/cifer/innerprod/fullysec/dmcfe.h

```diff
@@ -0,0 +1,118 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#ifndef CIFER_DMCFE_H
+#define CIFER_DMCFE_H
+
+#include "cifer/data/mat.h"
+#include "cifer/data/vec_curve.h"
+
+/**
+ * \file
+ * \ingroup fullysec
+ * \brief Decentralized multi-client inner-product scheme based on the paper
+ * "Decentralized Multi-Client Functional Encryption for Inner Product" by
+ * Chotard, Dufour Sans, Gay, Phan, and Pointcheval.
+ */
+
+/**
+ * cfe_dmcfe_client represents a client in a decentralized multi-client scheme
+ * scheme.
+ */
+typedef struct cfe_dmcfe_client {
+    size_t idx;
+    BIG_256_56 client_sec_key_big;
+    ECP_BN254 client_pub_key;
+    cfe_mat share;
+    cfe_vec s;
+    mpz_t order;
+    BIG_256_56 order_big;
+} cfe_dmcfe_client;
+
+/**
+ * Configures a new client for the dmcfe scheme.
+ *
+ * @param c A pointer to an uninitialized struct representing the scheme
+ * @param idx Identification value of the client; it is assumed that if there are
+ * n clients, their identifications are from [0,n)
+ */
+void cfe_dmcfe_client_init(cfe_dmcfe_client *c, size_t idx);
+
+/**
+ * Frees the memory occupied by the struct members. It does not free the
+ * memory occupied by the struct itself.
+ *
+ * @param c A pointer to an instance of the scheme (*initialized* cfe_dmcfe_client
+ * struct)
+ */
+void cfe_dmcfe_client_free(cfe_dmcfe_client *c);
+
+/**
+ * Sets a secret key share for the client based on the public keys all the
+ * participant. Note that the function assumes that if there is n clients,
+ * their identifications are from [0,n).
+ *
+ * @param c A pointer to an initialized struct representing the scheme
+ * @param pub_keys An array of public keys
+ * @param num_clients Number of clients
+ */
+void cfe_dmcfe_set_share(cfe_dmcfe_client *c, ECP_BN254 *pub_keys, size_t num_clients);
+
+/**
+ * Client c encrypts input value x, a coordinate of a vector. All clients should use
+ * the same label to encrypt a vector.
+ *
+ * @param cipher A pointer to ECP_BN254 struct (the resulting ciphertext will be
+ * stored here)
+ * @param x The input value
+ * @param label A string label of the encrypted vector
+ * @param c A pointer to an instance of the scheme (*initialized* cfe_dmcfe_client
+ * struct)
+ */
+void cfe_dmcfe_encrypt(ECP_BN254 *cipher, mpz_t x, char *label, cfe_dmcfe_client *c);
+
+/**
+ * Sets a share of a decryption key needed for the decryption of an inner product
+ * of encrypted vector and y.
+ *
+ * @param key_share A pointer to initialized cfe_vec_G2 of size 2 (the key share will
+ * be save here)
+ * @param y A pointer to the inner-product vector
+ * @param c A pointer to an initialized struct representing the scheme
+ */
+void cfe_dmcfe_generate_key_share(cfe_vec_G2 *key_share, cfe_vec *y, cfe_dmcfe_client *c);
```

Comment:
As far as I see, param `y` isn't a result of the function, so it can be after `c` in the param list.